lib/jsonapi/deserializable/resource.rb

```diff
@@ -1,110 +1,155 @@
       end
 
-      self.attr_blocks = {}
+      @class_cache = {}
+
+      self.configuration       = Configuration.new
+      self.attr_blocks         = {}
       self.has_one_rel_blocks  = {}
       self.has_many_rel_blocks = {}
 
       def self.inherited(klass)
         super
-        klass.type_block  = type_block
-        klass.id_block    = id_block
-        klass.attr_blocks = attr_blocks.dup
+        klass.configuration       = configuration.dup
+        klass.type_block          = type_block
+        klass.id_block            = id_block
+        klass.attr_blocks         = attr_blocks.dup
         klass.has_one_rel_blocks  = has_one_rel_blocks.dup
         klass.has_many_rel_blocks = has_many_rel_blocks.dup
       end
 
+      def self.configure
+        yield(configuration)
+      end
+
+      def self.[](name)
+        @class_cache[name] ||= Class.new(self)
+      end
+
       def self.call(payload)
         new(payload).to_h
       end
 
       def initialize(payload)
+        Parser::Resource.parse!(payload)
         @document = payload
         @data = @document['data']
         @type = @data['type']
         @id   = @data['id']
         @attributes    = @data['attributes'] || {}
         @relationships = @data['relationships'] || {}
         deserialize!
+        freeze
       end
 
       def to_hash
         @hash
       end
       alias to_h to_hash
 
+      attr_reader :reverse_mapping
+
       private
 
-      def deserialize!
-        @hash = {}
-        deserialize_type!
-        deserialize_id!
-        deserialize_attrs!
-        deserialize_rels!
+      def configuration
+        self.class.configuration
       end
 
-      def deserialize_type!
-        return unless @type && self.class.type_block
-        instance_exec(@type, &self.class.type_block)
+      def register_mappings(keys, path)
+        keys.each do |k|
+          @reverse_mapping[k] = path
+        end
       end
 
-      def deserialize_id!
-        return unless @id && self.class.id_block
-        instance_exec(@id, &self.class.id_block)
+      def deserialize!
+        @reverse_mapping = {}
+        hashes = [deserialize_type, deserialize_id,
+                  deserialize_attrs, deserialize_rels]
+        @hash = hashes.reduce({}, :merge)
       end
 
-      def deserialize_attrs!
-        self.class.attr_blocks.each do |attr, block|
-          next unless @attributes.key?(attr)
-          instance_exec(@attributes[attr], &block)
-        end
+      def deserialize_type
+        block = self.class.type_block || configuration.default_type
+        hash = block.call(@type)
+        register_mappings(hash.keys, '/data/type')
+        hash
       end
 
-      def deserialize_rels!
-        deserialize_has_one_rels!
-        deserialize_has_many_rels!
+      def deserialize_id
+        return {} unless @id
+        block = self.class.id_block || configuration.default_id
+        hash  = block.call(@id)
+        register_mappings(hash.keys, '/data/id')
```

Comment:
So, my understanding of the `register_mapping` code is to cache parsing results within methods like this, so we can go the reverse way.

I think this makes sense from a performance standpoint, but does clutter the code a bit. I'd consider making `reverse_mapping` a method, possibly implemented by a separate class. The scenarios I can think of using this, performance shouldn't matter much and it would make the code tighter IMO. Just a thought.